R/ui.R

```diff
@@ -103,17 +103,66 @@ ui_w <- function(text = "", env = parent.frame()) {
 #' # o Are you sure you did it right?
 #' }
 #' @noRd
-ui_throw <- function(message = "Internal error", details = character(0), env = parent.frame()) {
+ui_throw <- function(message = "Internal error", details = character(0),
+                     env = parent.frame(),
+                     glue_open = "{", glue_close = "}") {
   message <- cli_format_text(message, env = env)
 
-  if (length(details) != 0L) {
-    details <- glue::glue_collapse(details, sep = "\n")
-    message <- glue::glue(message, details, .sep = "\n")
-  }
+  error_messages <- purrr::map(
+    c(message, details),
+    glue,
+    .open = glue_open,
+    .close = glue_close
+  )
+
+  error_messages <- subset_lines_to_fit_in_limit(error_messages, 8000L)
+
+  message <- glue_collapse(error_messages, sep = "\n")
 
-  rlang::abort(message, class = "rextendr_error")
+  withr::with_options(
+    # Valid values are something between 1000 and 8170
+    # This will be set by {rlang} in the future release,
+    # https://github.com/r-lib/rlang/pull/1214
+    list(warning.length = 8000),
+    rlang::abort(message, class = "rextendr_error")
+  )
 }
 
 cli_format_text <- function(message, env = parent.frame()) {
   cli::cli_format_method(cli::cli_text(message, .envir = env))
 }
+
+#' Subset lines to fit within provided character limit.
+#'
+#' Verifies that `lines` fit within given limit of `max_length`.
+#' If not, lines are subset and an additional `truncation_notification`
+#' is appended to `lines`. The function ensures that the output message
+#' constructed by combining `lines` with `\n` separator
+#' fits within `max_length` limit.
+#'
+#' @param lines \[ ansi_character(n) \] Error messages.
+#' @param max_length Maximum total length of the collapsed message
+#' (measured using `nchar()`).
+#' @param truncation_notification Message pattern appended if any `lines`
+#' were removed. Understands `cli` interpolation and
+#' `n_removed` integer variable, which is set to the number of removed lines.
+#' @return \[ ansi_character(n) \] A subset of `lines` with
+#' `truncation_notification` appended (if needed).
+#' @noRd
+subset_lines_to_fit_in_limit <- function(lines,
+                                         max_length = 1000,
+                                         truncation_notification = "{.val {n_removed}} compiler messages not shown.") {
+  n_removed <- 100L
```

Comment:
```suggestion
```

It looks this line is unnecessary.

Reply:
This may be an overkill, but the trick here is to first compute the actual size of the message appended if lines are trimmed. This message is i) customizable ii) depends on the number of removed lines (which is unknown).
So I first assign an arbitrarily large value to `n_removed`, then I use `{cli}` to obtain the actual interpolated string and then use it to compute the length. Assuming we will never have more than 999 lines removed, it gives an upper limit on the length of the added string, which is enough to fit everything within the buffer.
If I remove this declaration, the `{cli}` will fail interpolation.

It is sort of a circular dependency -- `n_removed` depends on the length of the appended message, which in turn depends on `n_removed`.

Reply:
Ah, sorry, I just missed this variable is referenced by `truncation_notification`. Thanks for the explanation.

Reply:
I would suggest to add a few comments to the code explaining this reasoning. Whenever we have to do some weird trick to get something done, it's a good idea to write the reasoning down, as nobody will remember five years from now.